crates/matrix-sdk/src/authentication/oidc/http_client.rs

```diff
@@ -0,0 +1,191 @@
+    AsyncHttpClient, ErrorResponse, HttpClientError, HttpRequest, HttpResponse, RequestTokenError,
+};
+
+/// An HTTP client for making OAuth 2.0 requests.
+#[derive(Debug, Clone)]
+pub(super) struct OauthHttpClient {
+    pub(super) inner: reqwest::Client,
+    /// Rewrite HTTPS requests to use HTTP instead.
+    ///
+    /// This is a workaround to bypass some checks that require an HTTPS URL,
+    /// but we can only mock HTTP URLs.
+    #[cfg(test)]
+    pub(super) insecure_rewrite_https_to_http: bool,
+}
+
+impl<'c> AsyncHttpClient<'c> for OauthHttpClient {
+    type Error = HttpClientError<reqwest::Error>;
+
+    type Future =
+        Pin<Box<dyn Future<Output = Result<HttpResponse, Self::Error>> + Send + Sync + 'c>>;
+
+    fn call(&'c self, request: HttpRequest) -> Self::Future {
+        Box::pin(async move {
+            #[cfg(test)]
+            let request = if self.insecure_rewrite_https_to_http
+                && request.uri().scheme().is_some_and(|scheme| *scheme == http::uri::Scheme::HTTPS)
+            {
+                let mut request = request;
+
+                let mut uri_parts = request.uri().clone().into_parts();
+                uri_parts.scheme = Some(http::uri::Scheme::HTTP);
+                *request.uri_mut() = http::uri::Uri::from_parts(uri_parts)
+                    .expect("reconstructing URI from parts should work");
+
+                request
+            } else {
+                request
+            };
+
+            let response = self.inner.call(request).await?;
+
+            Ok(response)
+        })
+    }
+}
+
+/// Check the status code of the given HTTP response to identify errors.
+pub(super) fn check_http_response_status_code<T: ErrorResponse + 'static>(
+    http_response: &HttpResponse,
+) -> Result<(), RequestTokenError<HttpClientError<reqwest::Error>, T>> {
+    if http_response.status().as_u16() < 400 {
+        return Ok(());
+    }
+
+    let reason = http_response.body().as_slice();
+    let error = if reason.is_empty() {
+        RequestTokenError::Other("server returned an empty error response".to_owned())
+    } else {
+        match serde_json::from_slice(reason) {
+            Ok(error) => RequestTokenError::ServerResponse(error),
+            Err(error) => RequestTokenError::Other(error.to_string()),
+        }
+    };
+
+    Err(error)
+}
+
+/// Check that the server returned a response with a JSON `Content-Type`.
+pub(super) fn check_http_response_json_content_type<T: ErrorResponse + 'static>(
+    http_response: &HttpResponse,
+) -> Result<(), RequestTokenError<HttpClientError<reqwest::Error>, T>> {
+    let Some(content_type) = http_response.headers().get(http::header::CONTENT_TYPE) else {
+        return Ok(());
+    };
+
+    if content_type
+        .to_str()
+        // Check only the beginning of the content type, because there might be extra
+        // parameters, like a charset.
+        .is_ok_and(|ct| ct.to_lowercase().starts_with(mime::APPLICATION_JSON.essence_str()))
```

Comment:
A `Content-Type: application/jsonfoo` would pass this test. Thinking about the `;charset` is smart, but I don't think it's safe that way.

Reply:
I don't think that this is actually a safety issue. It's just a code path to error quickly if we didn't get the format that we expected.

Nothing prevents the server from announcing that they are sending JSON but send something else entirely. There is no point in making up a weird content-type to circumvent that check.